map/bookmark_manager.cpp

```diff
@@ -2517,9 +2541,16 @@ void BookmarkManager::CreateCategories(KMLDataCollection && dataCollection, bool
 
     UserMarkIdStorage::Instance().EnableSaving(false);
 
+    auto groupId = GetCategoryByFileName(fileName);
     // Set autoSave = false now to avoid useless saving in NotifyChanges().
     // autoSave flag will be assigned in the end of this function.
-    auto const groupId = CreateBookmarkCategory(std::move(categoryData), false /* autoSave */);
+    if (groupId != kml::kInvalidMarkGroupId)
+    {
+      isReloading = true;
```

Comment:
Woould it be clearer to split the existing confusing mix of creating and loading into separate cases?
```
// Despite the name, this method is called not only for newly created categories, but also for loaded/imported ones.
void BookmarkManager::CreateCategories(...)
```

Would it be easier to have `CreateCategory(one category data)` when user creates it in the app or imports a file, LoadCategories when they're loaded, and UpdateCategory when it's updated? Any thoughts?

Reply:
This method is enormous big and i am afraid of duplicating the code here.
Ill try to smth with that but I don't want to start the whole method refactoring in this PR (there are a lot of logic at one place):

```cpp
void BookmarkManager::CreateCategories(KMLDataCollection && dataCollection, bool autoSave /* = false */)
{
  CHECK_THREAD_CHECKER(m_threadChecker, ());
  kml::GroupIdSet loadedGroups;

  bool isReloading = false;
  for (auto const & [fileName, fileDataPtr] : dataCollection)
  {
    auto & fileData = *fileDataPtr;
    auto & categoryData = fileData.m_categoryData;

    // Initialize timestamp for newly created or imported categories.
    if (categoryData.m_lastModified == kml::Timestamp{})
      categoryData.m_lastModified = fileName.empty() ? kml::TimestampClock::now() : FileModificationTimestamp(fileName);

    if (!UserMarkIdStorage::Instance().CheckIds(fileData) || HasDuplicatedIds(fileData))
    {
      LOG(LINFO, ("Reset bookmark ids in the file", fileName));
      //TODO: notify subscribers(like search subsystem). This KML could have been indexed.
      ResetIds(fileData);
    }

    std::unordered_map<kml::CompilationId, BookmarkCategory *> compilations;
    std::unordered_set<std::string> compilationNames;
    for (auto & compilation : fileData.m_compilationsData)
    {
      SetUniqueName(compilation, [&compilationNames](auto const & name)
      {
        return compilationNames.count(name) == 0;
      });

      auto const compilationId = compilation.m_compilationId;
      auto childGroup = CreateBookmarkCompilation(std::move(compilation));
      categoryData.m_compilationIds.push_back(childGroup->GetID());

      compilations.emplace(compilationId, childGroup);
      compilationNames.emplace(childGroup->GetName());
      childGroup->SetFileName(fileName);
      childGroup->SetServerId(fileData.m_serverId);
    }

    SetUniqueName(categoryData, [this](auto const & name) { return !IsUsedCategoryName(name); });

    UserMarkIdStorage::Instance().EnableSaving(false);

    auto groupId = GetCategoryByFileName(fileName);
    // Set autoSave = false now to avoid useless saving in NotifyChanges().
    // autoSave flag will be assigned in the end of this function.
    if (groupId != kml::kInvalidMarkGroupId)
    {
      isReloading = true;
      ReloadBookmarkCategory(groupId, std::move(categoryData), false /* autoSave */);
    }
    else
      groupId = CreateBookmarkCategory(std::move(categoryData), false /* autoSave */);
    loadedGroups.insert(groupId);
    auto * group = GetBmCategory(groupId);
    group->SetFileName(fileName);
    group->SetServerId(fileData.m_serverId);

    // Restore sensitive info from the cache.
    auto const cacheIt = m_restoringCache.find(fileName);
    if (cacheIt != m_restoringCache.end() &&
        (group->GetServerId().empty() || group->GetServerId() == cacheIt->second.m_serverId) &&
        cacheIt->second.m_accessRules != group->GetCategoryData().m_accessRules)
    {
      group->SetServerId(cacheIt->second.m_serverId);
      group->SetAccessRules(cacheIt->second.m_accessRules);
      group->EnableAutoSave(autoSave);
    }

    for (auto const & [compilationId, compilation] : compilations)
    {
      UNUSED_VALUE(compilationId);
      compilation->SetParentId(groupId);
      auto const & catData = group->GetCategoryData();
      compilation->SetAccessRules(catData.m_accessRules);
      compilation->SetAuthor(catData.m_authorName, catData.m_authorId);
    }

    for (auto & bmData : fileData.m_bookmarksData)
    {
      auto const compilationIds = bmData.m_compilations;
      auto * bm = CreateBookmark(std::move(bmData));
      bm->Attach(groupId);
      group->m_userMarks.insert(bm->GetId());
      for (auto const c : compilationIds)
      {
        auto const it = compilations.find(c);
        if (it == compilations.end())
        {
          LOG(LERROR, ("Incorrect compilation id", c, "into", fileName));
          continue;
        }
        bm->AttachCompilation(it->second->GetID());
        it->second->AttachUserMark(bm->GetId());
      }
      m_changesTracker.OnAttachBookmark(bm->GetId(), groupId);
    }
    for (auto & trackData : fileData.m_tracksData)
    {
      auto track = std::make_unique<Track>(std::move(trackData), group->HasElevationProfile());
      auto * t = AddTrack(std::move(track));
      t->Attach(groupId);
      group->m_tracks.insert(t->GetId());
      if (t->IsInteractive())
        SetDefaultTrackSelection(t->GetId(), false /* showInfoSign */);
    }
    UpdateTrackMarksVisibility(groupId);
    UserMarkIdStorage::Instance().EnableSaving(true);
  }
  m_restoringCache.clear();

  // During the reloading process the file shouldn't be re-saved on disk because it should be already up to date.
  // In other case race condition may occur when multiple devices are used.
  NotifyChanges(!isReloading /* Save bookmarks on disk */);

  for (auto const & groupId : loadedGroups)
    GetBmCategory(groupId)->EnableAutoSave(autoSave);
}
```